### PR TITLE
MVCC

### PR DESCRIPTION
**Pull request by zfrenette**
## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                secondaryIndexesWrapper.clear()
+        );
+    }
+
+    /**
+     * Returns a set of all keys present.
+     * @return A set containing all keys present in this ImmutableCorfuTable.
+     */
+    public java.util.Set<K> keySet() {
+        // TODO: Can this call to toJavaSet() be avoided?
```
[the remark this answers is not in the document]
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java]: I have removed this ToDo since keySet() is not a commonly used API.

File: runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+     * @param indexKey The index key used to query the secondary index
+     * @param <I> The type of the index key.
+     * @return A collection of map entries satisfying this index query.
+     */
+    public <I> Collection<java.util.Map.Entry<K, V>> getByIndex(@Nonnull final Index.Name indexName, I indexKey) {
+        final String secondaryIndexName = indexName.get();
+        Option<Map<K, V>> optMap = secondaryIndexesWrapper.contains(secondaryIndexName, indexKey);
+
+        if (!optMap.isEmpty()) {
+            // TODO: Can this call to toJavaMap() be avoided?
```
[the remark this answers is not in the document]
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java]: I have implemented this ToDo by updating the API to return `Iterable`. Please have a look.

File: infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java [hunk 3]
```diff
@@ -43,7 +43,7 @@ public void testPollRound() {
         );
         Duration time = Duration.ofMillis(System.currentTimeMillis() - start);
         assertThat(time).isGreaterThan(Duration.ofMillis(450));
-        assertThat(time).isLessThan(Duration.ofSeconds(2));
+        assertThat(time).isLessThan(Duration.ofSeconds(10));
```
Comment (Maithem) [inline infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java:46]: Not sure why this is needed. If it is for test stability then I would suggest its done in a separate PR.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 4]
```diff
@@ -46,7 +46,7 @@
  *  TODO: Generalize to all SMR objects.
  */
 @Slf4j
-public class CheckpointWriter<T extends StreamingMap> {
+public class CheckpointWriter<T extends ICorfuTable<?, ?>> {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java:49]: Why are these wild cards needed?

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 5]
```diff
@@ -230,9 +231,14 @@ public static class CorfuRuntimeParameters extends RuntimeParameters {
...
-         * The maximum number of entries in the cache.
+         * The maximum number of entries in the AddressSpaceView cache.
          */
-        long maxCacheEntries;
+        long maxCacheEntries = 5000;
+
+        /*
+         * The maximum number of entries in the MVOCache.
+         */
+        long maxMvoCacheEntries = 5000;
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:241]: I would tune these defaults down, such that the total  of maxCacheEntries + maxMvoCacheEntries equals the old default. The default behavior should not consume more memory than previous releases.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 6]
```diff
@@ -248,9 +254,14 @@ public static class CorfuRuntimeParameters extends RuntimeParameters {
...
         /*
-         * Sets expireAfterAccess and expireAfterWrite in seconds.
+         * Sets expireAfterAccess and expireAfterWrite for the AddressSpaceView cache in seconds.
          */
         long cacheExpiryTime = Long.MAX_VALUE;
+
+        /*
+         * Sets expireAfterAccess and expireAfterWrite for the MVOCache in seconds.
+         */
+        long mvoCacheExpiryTime = 300;
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:264]: I don't think we have enough data to support this optimization and for the sake of simplicity I would suggest just letting the LRU doing the eviction.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 7]
```diff
@@ -426,6 +439,7 @@ public static class CorfuRuntimeParametersBuilder extends RuntimeParametersBuild
             private int streamBatchSize = 10;
             private int checkpointReadBatchSize = 5;
             private Duration runtimeGCPeriod = Duration.ofMinutes(20);
+            private Duration mvoAutoSyncPeriod = Duration.ofMinutes(5);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:442]: It doesnt seem like this config is being used.

File: runtime/src/main/java/org/corfudb/runtime/DistributedCheckpointer.java [hunk 8]
```diff
@@ -6,10 +6,7 @@
 import org.corfudb.runtime.CorfuCompactorManagement.CheckpointingStatus;
 import org.corfudb.runtime.CorfuCompactorManagement.CheckpointingStatus.StatusType;
 import org.corfudb.runtime.CorfuStoreMetadata.TableName;
-import org.corfudb.runtime.collections.CorfuStore;
-import org.corfudb.runtime.collections.StreamingMap;
-import org.corfudb.runtime.collections.Table;
-import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.collections.*;
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/DistributedCheckpointer.java:9]: Hmm, need an auto linter. Expand these imports.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java [hunk 9]
```diff
@@ -382,9 +384,12 @@ public boolean equals(Object o){
     public String toString(){
         StringBuilder stringBuilder = new StringBuilder(corfuTable.size());
         stringBuilder.append("{");
-        for (Map.Entry<CorfuRecordId, ByteString> entry : corfuTable.entrySet()) {
+
+        for (Iterator<Map.Entry<CorfuRecordId, ByteString>> it = corfuTable.entryStream().iterator(); it.hasNext(); ) {
+            Map.Entry<CorfuRecordId, ByteString> entry = it.next();
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java:389]: This might turn out to be expensive, does this dump the binary payload? I dont think we should do that for logging purposes.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 10]
```diff
@@ -624,6 +638,18 @@ public CorfuRuntimeParameters.CorfuRuntimeParametersBuilder cacheDisabled(boolea
...
                 this.maxCacheEntries = maxCacheEntries;
+
+                final long maxCacheSizeThresh = 10000L;
+                final long largeMvoCacheSize = 25000L;
+                if (maxCacheEntries >= maxCacheSizeThresh) {
+                    this.maxMvoCacheEntries = largeMvoCacheSize;
+                }
+
+                return this;
+            }
```
[the remark this answers is not in the document]
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:649]: Right - This was just done for testing on PKS. It will be removed in the next push.

File: runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+     * @param indexKey The index key used to query the secondary index
+     * @param <I> The type of the index key.
+     * @return An Iterable of map entries satisfying this index query.
+     */
+    public <I> Iterable<java.util.Map.Entry<K, V>> getByIndex(@Nonnull final Index.Name indexName, I indexKey) {
+        return secondaryIndexesWrapper.contains(indexName.get(), indexKey)
+                .<Iterable<java.util.Map.Entry<K, V>>>map(IterableWrapper::new)
+                .getOrElse(Collections.emptySet());
+    }
+
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java:169]: I know you haven't introduced this, but I think we should not allow for null-queries. 

Since multiple values can map to null, It is ambiguous and I don't think application code should use null for queries. Preventing null from the secondary index will also save. No null index = less memory usage.

File: runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+    private class IterableWrapper implements Iterable<java.util.Map.Entry<K, V>> {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java:170]: I would put this in a separate class. It seems like a common utility for efficient conversion between vavr types to Java's Map types.

File: runtime/src/main/java/org/corfudb/runtime/collections/Table.java [hunk 13]
```diff
@@ -42,7 +46,22 @@
...
+    // can acquire the VLO lock and cause the other 3 threads to wait, but after acquiring the VLO lock, the thread
+    // gets block on parallel stream, because the pool is exhausted with threads that are trying to acquire the VLO
+    // look, which creates a circular dependency. In other words, a deadlock.
+
+    protected static final ForkJoinPool pool = new ForkJoinPool(Math.max(Runtime.getRuntime().availableProcessors() / 2, 1),
+            pool -> {
+                final ForkJoinWorkerThread worker = ForkJoinPool.defaultForkJoinWorkerThreadFactory.newThread(pool);
+                worker.setName("Table-Forkjoin-pool-" + worker.getPoolIndex());
+                return worker;
+            }, null, true);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/Table.java:62]: Can you change this to a fixed thread pool and make the number of workers configurable via the Runtime options?

The problem with Forkjoin pool is that it can fork more threads than specified if some of the work is blocking. The scan and filter shouldn't have blocking code. Having a fixed thread pool would make this issue easier to observe in a profiler.

**Review by Maithem — APPROVED** [review]
Great patch!

**Review by Lujie1996 — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java (hunk 5 above)
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: Done. I have divided them evenly at 2500 entries.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java (hunk 6 above)
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: Removed.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java (hunk 7 above)
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: Removed. This was a left over from previous code.

File: runtime/src/main/java/org/corfudb/runtime/DistributedCheckpointer.java (hunk 8 above)
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/DistributedCheckpointer.java]: Expanded.

File: runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java (hunk 12 above)
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java]: Done - I have extracted these two classes.

File: runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java (hunk 11 above)
Comment (zfrenette) [reply runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java]: I want to maintain the same semantics for both CorfuTable implementations. I can address this in a subsequent PR.
